File: proxy_checker.js
const fs = require("fs");
const fetch = require("node-fetch");
const { HttpsProxyAgent } = require("https-proxy-agent");
const { SocksProxyAgent } = require("socks-proxy-agent");

function loadConfig(configPath = "config.json") {
    try {
        const configData = fs.readFileSync(configPath, "utf8");
        return JSON.parse(configData);
    } catch (error) {
        console.error(`Error loading configuration: ${error.message}`);
        return {
            proxy_file: "proxy.txt",
            output_file: "working_proxies.txt",
            test_url: "http://www.google.com",
            timeout: 5,
            concurrent_checks: 20,
            save_to_input_file: false,
        };
    }
}

function parseProxyString(proxyStr) {
    let protocol = "http";
    let username = "";
    let password = "";
    let host = "";
    let port = "";

    // Перевіряємо, чи є протокол
    if (proxyStr.includes("://")) {
        [protocol, proxyStr] = proxyStr.split("://");
    }

    // Якщо є @ - значить є автентифікація
    if (proxyStr.includes("@")) {
        const [auth, hostPort] = proxyStr.split("@");
        [username, password] = auth.split(":");
        [host, port] = hostPort.split(":");
    } else if (proxyStr.includes(":")) {
        // Формат без @ але з автентифікацією (ip:port:username:password)
        const parts = proxyStr.split(":");
        if (parts.length === 4) {
            [host, port, username, password] = parts;
        } else {
            [host, port] = parts;
        }
    }

    return {
        protocol: protocol.toLowerCase(),
        host,
        port: parseInt(port),
        username,
        password,
        toString() {
            const auth = username && password ? `${username}:${password}@` : "";
            return `${protocol}://${auth}${host}:${port}`;
        },
    };
}

function parseArgs() {
    const args = process.argv.slice(2);
    let configPath = "config.json";

    for (let i = 0; i < args.length; i++) {
        if ((args[i] === "-c" || args[i] === "--config") && i + 1 < args.length) {
            configPath = args[i + 1];
            break;
        }
    }

    return { configPath };
}

const { configPath } = parseArgs();
console.log(`Using configuration file: ${configPath}`);

const config = loadConfig(configPath);

const PROXY_FILE = config.proxy_file;
const OUTPUT_FILE = config.output_file;
const TEST_URL = config.test_url;
const TIMEOUT = config.timeout * 1000;
const CONCURRENT_CHECKS = config.concurrent_checks;
const SAVE_TO_INPUT_FILE = config.save_to_input_file;

function readProxiesFromFile(filePath) {
    try {
        const data = fs.readFileSync(filePath, "utf8");
        return data.split("\n").filter((line) => line.trim() !== "");
    } catch (error) {
        console.error(`Error reading a file:${error.message}`);
        return [];
    }
}

async function checkHttpProxy(proxyInfo) {
    return new Promise(async (resolve) => {
        try {
            const proxyUrl = proxyInfo.toString();
            const agent = new HttpsProxyAgent(proxyUrl);

            const response = await fetch(TEST_URL, {
                agent,
                timeout: TIMEOUT,
                method: "HEAD",
            });

            resolve(response.ok);
        } catch (error) {
            resolve(false);
        }
    });
}

async function checkSocksProxy(proxyInfo) {
    return new Promise(async (resolve) => {
        try {
            const proxyUrl = proxyInfo.toString();
            const agent = new SocksProxyAgent(proxyUrl);

            const response = await fetch(TEST_URL, {
                agent,
                timeout: TIMEOUT,
                method: "HEAD",
            });

            resolve(response.ok);
        } catch (error) {
            resolve(false);
        }
    });
}

async function checkProxy(proxyStr) {
    try {
        const proxyInfo = parseProxyString(proxyStr);

        switch (proxyInfo.protocol) {
            case "http":
            case "https":
                return await checkHttpProxy(proxyInfo);
            case "socks4":
            case "socks5":
                return await checkSocksProxy(proxyInfo);
            default:

                proxyInfo.protocol = "http";
                return await checkHttpProxy(proxyInfo);
        }
    } catch (error) {
        return false;
    }
}

async function processProxiesInBatches(proxies) {
    const workingProxies = [];
    const totalProxies = proxies.length;
    let processedCount = 0;

    console.log(`Starting proxy check of ${totalProxies} proxies...`);

    for (let i = 0; i < totalProxies; i += CONCURRENT_CHECKS) {
        const batch = proxies.slice(i, i + CONCURRENT_CHECKS);
        const promises = batch.map(async (proxy) => {
            const isWorking = await checkProxy(proxy);
            processedCount++;

            if (processedCount % 10 === 0 || processedCount === totalProxies) {
                console.log(
                    `Progress: ${processedCount}/${totalProxies} (${Math.round((processedCount / totalProxies) * 100)}%)`
                );
            }

            if (isWorking) {
                console.log(`✅ Works: ${proxy}`);
                workingProxies.push(proxy);
            } else {
                console.log(`❌ Not works: ${proxy}`);
            }

            return { proxy, isWorking };
        });

        await Promise.all(promises);
    }

    return workingProxies;
}

function saveWorkingProxies(proxies, filePath) {
    try {
        fs.writeFileSync(filePath, proxies.join("\n"));
        console.log(`Saved ${proxies.length} working proxies to file ${filePath}`);
    } catch (error) {
        console.error(`Error saving file: ${error.message}`);
    }
}

async function main() {
    console.log("Starting proxy check...");
    console.log(`Using configuration from file: ${configPath}`);
    console.log(`Proxy file: ${PROXY_FILE}`);
    console.log(`Output file: ${OUTPUT_FILE}`);
    console.log(`URL for testing: ${TEST_URL}`);
    console.log(`Timeout: ${TIMEOUT / 1000} seconds`);
    console.log(`Number of concurrent checks: ${CONCURRENT_CHECKS}`);
    console.log(`Save to input file: ${SAVE_TO_INPUT_FILE ? "Yes" : "No"}`);

    const proxies = readProxiesFromFile(PROXY_FILE);
    console.log(`Loaded ${proxies.length} proxies from file ${PROXY_FILE}`);

    if (proxies.length === 0) {
        console.error("No proxies found for checking.");
        return;
    }

    const workingProxies = await processProxiesInBatches(proxies);

    console.log(`\nResults of the check:`);
    console.log(`Total proxies: ${proxies.length}`);
    console.log(`Working proxies: ${workingProxies.length}`);
    console.log(`Not working proxies: ${proxies.length - workingProxies.length}`);

    saveWorkingProxies(workingProxies, OUTPUT_FILE);

    if (SAVE_TO_INPUT_FILE) {
        console.log(`Saving working proxies to input file: ${PROXY_FILE}`);
        saveWorkingProxies(workingProxies, PROXY_FILE);
    } else {
        console.log(`Not saving to input file (disabled in configuration)`);
    }
}

main().catch((error) => {
    console.error("Error executing the program:", error);
});

File: proxy_checker.py
import os
import sys
import time
import asyncio
import aiohttp
import aiofiles
import json
import argparse
from urllib.parse import urlparse
from aiohttp_socks import ProxyConnector, ProxyType

DEFAULT_CONFIG_FILE = 'config.json'

def parse_args():
    parser = argparse.ArgumentParser(description='Proxy Checker - checking the proxy')
    parser.add_argument('-c', '--config', default=DEFAULT_CONFIG_FILE, 
                        help=f'Path to the configuration file (default: {DEFAULT_CONFIG_FILE})')
    return parser.parse_args()

def load_config(config_file):
    try:
        with open(config_file, 'r') as file:
            return json.load(file)
    except Exception as e:
        print(f"Error loading config file: {e}")
        return {
            "proxy_file": "proxy.txt",
            "output_file": "working_proxies.txt",
            "test_url": "http://www.google.com",
            "timeout": 5,
            "concurrent_checks": 20,
            "save_to_input_file": False
        }

def save_config(config_data, config_file):
    try:
        with open(config_file, 'w') as file:
            json.dump(config_data, file, indent=4)
        print(f"Configuration saved to {config_file}")
        return True
    except Exception as e:
        print(f"Error saving config file: {e}")
        return False

args = parse_args()
CONFIG_FILE = args.config

config = load_config(CONFIG_FILE)
PROXY_FILE = config["proxy_file"]
OUTPUT_FILE = config["output_file"]
TEST_URL = config["test_url"]
TIMEOUT = config["timeout"]
CONCURRENT_CHECKS = config["concurrent_checks"]
SAVE_TO_INPUT_FILE = config.get("save_to_input_file", False)

def read_proxies_from_file(file_path):
    try:
        with open(file_path, 'r') as file:
            return [line.strip() for line in file if line.strip()]
    except Exception as e:
        print(f"Error reading file: {e}")
        return []

async def check_proxy(proxy_url):
    try:
        parsed = urlparse(proxy_url)
        proxy_type = parsed.scheme
        host = parsed.hostname
        port = parsed.port
        
        if proxy_type == 'http' or proxy_type == 'https':
            connector = ProxyConnector.from_url(proxy_url)
        elif proxy_type == 'socks4':
            connector = ProxyConnector(
                proxy_type=ProxyType.SOCKS4,
                host=host,
                port=port
            )
        elif proxy_type == 'socks5':
            connector = ProxyConnector(
                proxy_type=ProxyType.SOCKS5,
                host=host,
                port=port
            )
        else:
            connector = ProxyConnector.from_url("http://" + proxy_url)
        
        timeout = aiohttp.ClientTimeout(total=TIMEOUT)
        async with aiohttp.ClientSession(connector=connector, timeout=timeout) as session:
            async with session.get(TEST_URL) as response:
                return response.status >= 200 and response.status < 300
    except:
        return False

async def process_proxies(proxies):
    working_proxies = []
    total_proxies = len(proxies)
    processed_count = 0
    
    print(f"Starting proxy check of {total_proxies} proxies...")
    
    for i in range(0, total_proxies, CONCURRENT_CHECKS):
        batch = proxies[i:i + CONCURRENT_CHECKS]
        tasks = [check_proxy(proxy) for proxy in batch]
        
        results = await asyncio.gather(*tasks, return_exceptions=True)
        
        for j, result in enumerate(results):
            proxy = batch[j]
            is_working = isinstance(result, bool) and result
            
            processed_count += 1
            
            if processed_count % 10 == 0 or processed_count == total_proxies:
                print(f"Progress: {processed_count}/{total_proxies} ({round(processed_count/total_proxies*100)}%)")
            
            if is_working:
                print(f"✅ It works: {proxy}")
                working_proxies.append(proxy)
            else:
                print(f"❌ Not working: {proxy}")
    
    return working_proxies

async def save_working_proxies(proxies, file_path):
    try:
        async with aiofiles.open(file_path, 'w') as file:
            await file.write('\n'.join(proxies))
        print(f"Saved {len(proxies)} working proxies to file {file_path}")
    except Exception as e:
        print(f"Error saving file: {e}")

async def main():
    print('Starting proxy check...')
    print(f'Using configuration from {CONFIG_FILE}')
    print(f'Proxy file: {PROXY_FILE}')
    print(f'Output file: {OUTPUT_FILE}')
    print(f'Test URL: {TEST_URL}')
    print(f'Timeout: {TIMEOUT} seconds')
    print(f'Concurrent checks: {CONCURRENT_CHECKS}')
    print(f'Save to input file: {SAVE_TO_INPUT_FILE}')
    
    proxies = read_proxies_from_file(PROXY_FILE)
    print(f"Loaded {len(proxies)} proxies from file {PROXY_FILE}")
    
    if not proxies:
        print('No proxies found for checking.')
        return
    
    working_proxies = await process_proxies(proxies)
    
    print(f"\nResults of the check:")
    print(f"Total proxies: {len(proxies)}")
    print(f"Working proxies: {len(working_proxies)}")
    print(f"Not working proxies: {len(proxies) - len(working_proxies)}")
    
    await save_working_proxies(working_proxies, OUTPUT_FILE)
    
    if SAVE_TO_INPUT_FILE:
        print(f"Saving working proxies back to input file {PROXY_FILE}")
        await save_working_proxies(working_proxies, PROXY_FILE)
    else:
        print(f"Not saving to input file (disabled in config)")

if __name__ == "__main__":
    if sys.platform.startswith('win'):
        asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
    asyncio.run(main()) 